libhif/hif-source.c

```diff
@@ -1657,22 +1657,53 @@ hif_source_download_package (HifSource *source,
 			     HifState *state,
 			     GError **error)
 {
+	_cleanup_ptrarray_unref_ GPtrArray *packages = g_ptr_array_new ();
+	_cleanup_free_ gchar *basename = NULL;
+
+	g_ptr_array_add (packages, pkg);
+
+	if (!hif_source_download_packages (source, packages, directory, state, error))
+		return NULL;
+
+	/* build return value */
+	basename = g_path_get_basename (hy_package_get_location (pkg));
+	return g_build_filename (directory, basename, NULL);
+}
+
+/**
+ * hif_source_download_packages:
+ * @source: a #HifSource instance.
+ * @packages: (element-type HyPackage): Array of packages, must be from this source
+ * @directory: the destination directory.
+ * @state: a #HifState.
+ * @error: a #GError or %NULL..
+ *
+ * Downloads multiple packages from a source.  The target filename
+ * will be equivalent to `g_path_get_basename (hy_package_get_location
+ * (pkg))`.
+ **/
+gboolean
+hif_source_download_packages (HifSource *source,
+			      GPtrArray *packages,
```

Comment:
We need to have a design meeting about progress callbacks.  I think we need the ability to expose basically everything librepo supports, so clients can show things like total amount that will be downloaded (based on the repodata) etc.

This gets tricky because there's no good way right now to expose `LrHandle*` from librepo's python bindings such that it will be consumed by something using gobject-introspection.  g-i does have the concept of "foreign" (which is used for Cairo), but right now pygobject is special-casing Cairo, and I don't think we can (should) add in librepo to pygobject.

Maybe we can wrap it for just high level ops for things like that DNF API?